DOCS-3695, DOCS-3401: Add to dependency, validate, and reconfigure info (#4089)
viamrobotics/docs: docs/operate/get-started/other-hardware/_index.md

1. **Edit the `reconfigure` function** to do the following:

   - Get any values from the `config` object that the user has configured.

>    - Assign any default values as necessary to any optional attributes if the user hasn't configured them.

Author: ```suggestion
   - Assign default values as necessary to any optional attributes if the user hasn't configured them.
```

>    - If your module has dependencies, get the dependencies from the `dependencies` map and cast each resource according to which API it implements, as in [this <file>ackermann.py</file> example](https://github.com/mcvella/viam-ackermann-base/blob/main/src/ackermann.py).

Reviewer: Maybe show it once instead?
```
from typing import cast

        motors = config.attributes.fields["drive_motors"].list_value
        for m in motors:
            actual_motor = dependencies[Motor.get_resource_name(m)]
            motor = cast(Motor, actual_motor)
```

Author: Tried this in staging and I sort of think it takes up too much space and doesn't have enough context around it, esp. since not everyone will have dependencies.

>    When the user changes the configuration, the `reconfigure` function is called.

Reviewer 2: [optional nit] maybe make this the first sentence so we start by explaining what the function is doing

       </details><br>

1. **Edit the `reconfigure` function**, which gets called when the user changes the configuration.
   This function should do the following:

>    - Get any values from the `config` object that the user has configured.

Author: ```suggestion
   - If you assigned any configuration attributes to global variables, get the values from the latest `config` object and update the values of the global variables.
```

       </details><br>

   If you do not implement a `Reconfigure` function (see next step), your `Validate` function should also do the following:

   - Get any values from the `config` object that the user has configured.

>    - Assign default values as necessary to any optional attributes if the user hasn't configured them.<br><br>

Reviewer 2: agreed. If reconfigure is unimplemented, I think then the config structure needs to have `resource.AlwaysRebuild`, and all required attributes should be assigned in the constructor

>    - Assign default values as necessary to any optional attributes if the user hasn't configured them.<br><br>

Reviewer 2: I think validate should only be used to confirm that the config is formatted as expected?

>    - Assign default values as necessary to any optional attributes if the user hasn't configured them.<br><br>

Reviewer 3: Yeah agreed. I think we should recommend to implement Reconfigure if needing to do these things instead of putting into Validate.

The rest of the changes LGTM! thanks for improving this.

>    - Assign default values as necessary to any optional attributes if the user hasn't configured them.<br><br>

Reviewer 4: Michael's not wrong, but I prefer Martha's approach of using `AlwaysRebuild` as the default, and only writing `Reconfigure` if you absolutely need it. I kinda want to change the `viam module generate` template to create an `AlwaysRebuild` module with a comment saying that if you really need `Reconfigure()` you can comment out the `AlwaysRebuild` line. but that's a discussion for a different PR, not this one.

>    - Assign default values as necessary to any optional attributes if the user hasn't configured them.<br><br>

Author: So, for Go the generator does actually now put AlwaysRebuild in by default, and it doesn't even have a reconfigure function in there at all.

Okay I'll move this stuff to the constructor for Go. But for Python above, should the same pattern of be used? i.e., do more in the constructor (`new`) instead of in `validate_config`? Based on the code comments generated by the generator, I think no, and what I have here for Python is correct. Is that right?

>    - Assign default values as necessary to any optional attributes if the user hasn't configured them.<br><br>

Reviewer 4: (Sorry for my slow response) Yeah, I think that's right!

>    - Assign default values as necessary to any optional attributes if the user hasn't configured them.<br><br>

Reviewer 3: The more I think about it, the recommendation to put where varies a lot depending on the language used. On the Go side there are better utilities for managing the behavior for modules, which reduces the need for reconfigure.  

In python though I don't recommend excluding reconfigure ever. Unless something has changed I'm pretty sure you always need it.

It feels complicated to accomplish and maintain but: maybe the docs should differ by language?

>    - Assign default values as necessary to any optional attributes if the user hasn't configured them.<br><br>

Reviewer 3: Maybe someone on netcode or SDK could help inform better. @cheukt?

One caveat I've thinking of is explicit and implicit dependencies. When you want to have implicit dependencies, I'm pretty sure you need to have both validate and reconfigure: You have to return the implicit dependencies as a response to validate, so that you can get access to them in reconfigure.

>    - Assign default values as necessary to any optional attributes if the user hasn't configured them.<br><br>

Author: So @michaellee1019 the docs currently do differ by language and will continue to do so: https://deploy-preview-4089--viam-docs.netlify.app/operate/get-started/other-hardware/#implement-the-component-api --this is fine.
We are all on the same page that Python authors need to implement reconfigure because there's no easy way to alwaysRebuild with Python.

My question (sorry if it wasn't clear) was/is whether we should tell Python authors to edit the `new` class method to make it do anything at all beyond setting the name and calling reconfigure. The generator generates this for Python's constructor:
```
    @classmethod
    def new(
        cls, config: ComponentConfig, dependencies: Mapping[ResourceName, ResourceBase]
    ) -> Self:
        """This method creates a new instance of this Sensor component.
        The default implementation sets the name from the `config` parameter and then calls `reconfigure`.

        Args:
            config (ComponentConfig): The configuration for this resource
            dependencies (Mapping[ResourceName, ResourceBase]): The dependencies (both implicit and explicit)

        Returns:
            Self: The resource
        """
        return super().new(config, dependencies)
```
        
I think the answer is *no* changes here, the Python vs Go instructions should continue to differ not only in what reconfigure does/doesn't do but also in what the constructor does/doesn't do.

>    - Assign default values as necessary to any optional attributes if the user hasn't configured them.<br><br>

Author: To the point about implicit dependencies, I don't think that'd change anything in this PR would it, since we already say for both Go and Python that people need to get implicit deps in validate.

>    - Assign default values as necessary to any optional attributes if the user hasn't configured them.<br><br>

Author: Sorry thread is getting long.
For Go I will change it so these parts are in the constructor: 
   - Get any values from the `config` object that the user has configured.
   - Assign default values as necessary to any optional attributes if the user hasn't configured them.

For Python, I will keep these^ in validate as they are (unless someone says otherwise).

And then, LMK if I need to tell Go authors with implicit dependencies to implement reconfigure--I'm thinking I don't need to because as long as implicit deps are in validate, when the module rebuilds it should get them right? Or no?

>    - Assign default values as necessary to any optional attributes if the user hasn't configured them.<br><br>

Reviewer 4: My instinct is to do those two bullet points in the constructor or during reconfigure, and in `validate` merely check that the config appears complete. but I don't feel strongly and I'm content leaving it as-is. This all still LGTM

>    - Assign default values as necessary to any optional attributes if the user hasn't configured them.<br><br>

Author: I guess the first bullet, get any config values, doesn't need to be done in any of these functions since people can just do for example `imgFile, err := os.Open(s.cfg.ImagePath)` within the actual API function implementation instead of passing around global variables. So that "Get any values..." bullet becomes "If you assigned any config values to global variables, reassign them."?

>    - Assign default values as necessary to any optional attributes if the user hasn't configured them.<br><br>

Reviewer 4: Yeah, I like that!

>    - Assign default values as necessary to any optional attributes if the user hasn't configured them.<br><br>

Reviewer 3: Ok, sorry for the more philosophical comments above. I think that they apply to this but the docs are correct as they are in this PR. Just didn't see it earlier. It LGTM!

>    - Assign default values as necessary to any optional attributes if the user hasn't configured them.<br><br>

Author: Philosophy can be necessary when we are deciding how to guide users! I appreciate the discussion.

>    - Assign any default values as necessary to any optional attributes if the user hasn't configured them.<br><br>

Author: ```suggestion
   - Assign default values as necessary to any optional attributes if the user hasn't configured them.<br><br>
```

    cancelCtx:  cancelCtx,
    cancelFunc: cancelFunc,
  }

>   s.exampleValue = "default value"

Author: ```suggestion
  if cfg.ExampleValue != "" {
    s.exampleValue = "default value"
  }
```
